[20기 최준식 전현수] 회원가입 기능 구현

JunsikChoi opened the pull request:
## 수정 사항 간략한 한줄 요약
- SignUp View 추가
- 패스워드 / 이메일 검증 로직 추가
- 패스워드 해시 저장 기능 추가

- 예외 처리
    - JSON_DECODE_ERROR
    - INVALID_DATA_ERROR
    - KEY_ERROR
    - DUPLICATED_ENTRY_ERROR
    
## 수정 사항들 자세한 내용

### SignUp View
URL: /users/signup

- Input Key: email, password, phone_number, name, birthday(optional), gender(optional)
- output Key: status, message, user(성공시)

#### 200 SUCCESS
- SignUp 성공, 등록된 User 반환 (패스워드 제외)

#### 400 JSON_DECODE_ERROR
- 유효하지 않은 json 입력시

#### 400 INVALID_DATA_ERROR
- password | email | gender 필드 검증 실패시 

####  400 KEY_ERROR
- 필수 필드 존재하지 않을 시

#### 409 DUPLICATED_ENTRY_ERROR
- 중복 불가 필드에 중복 입력 요청시

## 체크 리스트
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

JunsikChoi committed:
MODIFY: Add Users app in settings.py

JunsikChoi committed:
MODIFY: Add Users URL in Root URL Conf

JunsikChoi committed:
CREATE: Add users/urls.py

JunsikChoi committed:
CREATE: Add utils and validators.py

JunsikChoi committed:
Add: SignUp View

JunsikChoi committed:
MODIFY: Add bcrypt

JunsikChoi committed:
MODIFY: Add SignUp Url Route

Kang-kyunghun commented on users/views.py:
```diff
@@ -0,0 +1,48 @@
+from utils.validators       import validate_email, validate_password, validate_gender, validate_duplicate, DuplicatedEntryError
+
+class SignUpView(View):
+
```
class 시작할 때 공백은 제거해 주세요.

Kang-kyunghun commented on users/views.py:
```diff
@@ -0,0 +1,48 @@
+class SignUpView(View):
+
+    def post(self, request):
+
```
함수 시작 할 때도 공백 제거해 주세요.

Kang-kyunghun commented on users/views.py:
```diff
@@ -0,0 +1,48 @@
+            hashed_password = bcrypt.hashpw(str(data.get('password')).encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
+
+            user = User.objects.create(
+                email           = data.get('email'),
+                password        = hashed_password,
+                phone_number    = data['phone_number'],
+                name            = data['name'],
+                birthday        = data.get('birthday'),
+                gender          = data.get('gender'),
```
어떤 data는 key로 바로 가져오고 어떤 data는 .get으로 가져오는데 그 이유가 있나요? 이미 위에서 validate로 데이터 가공은 끝난 것으로 보이는데 구별한 이유가 있나 궁금합니다.

JunsikChoi committed:
MODIFY: User Validator and SignUpView

- User Validator 부분 User app의 utils.py로 분리
- email, password validator가 검증된 값 리턴하게 수정, 가독성 고려
- validate_duplicate 함수 가독성 향상
- 그 외 코드포매팅

JunsikChoi committed:
FIX: Resolve Conflicts

JunsikChoi commented:
수정 요청 반영 및 Merge Conflict 해결했습니다.

JunsikChoi committed:
FIX: User URL Path

JunsikChoi committed:
MODIFY: JsonResponse Data Structure

JunsikChoi commented:
빼먹은 수정사항 반영 - url 구조 에러 수정, Json 응답 구조에서 부가 데이터 있는 경우 data field 하위로 통일했습니다.

JunsikChoi committed:
MODIFY: Change Duplicate Validator

- make validator not check fields in related table

Kang-kyunghun commented:
마이너한 코드 컨벤션만 고치면 될 것 같습니다.

JunsikChoi committed:
FIX: 마이너 코드 포매팅 수정

JunsikChoi committed:
MODIFY: Catch DataError Exception in SignUpView

JunsikChoi committed:
FIX: 마이너 코딩 컨벤션

JunsikChoi commented:
수정 완료했습니다!!!!!!!

JunsikChoi committed:
FIX: 마이너 컨벤션 수정

Kang-kyunghun commented:
LGTM
